File: src/test/java/com/procurement/clarification/service/EnquiryServiceImplTest.java
package com.procurement.clarification.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.procurement.clarification.exception.ErrorInsertException;
import com.procurement.clarification.model.dto.AuthorDto;
import com.procurement.clarification.model.dto.DataDto;
import com.procurement.clarification.model.dto.EnquiryDto;
import com.procurement.clarification.model.entity.EnquiryEntity;
import com.procurement.clarification.model.entity.EnquiryPeriodEntity;
import com.procurement.clarification.repository.EnquiryPeriodRepository;
import com.procurement.clarification.repository.EnquiryRepository;
import com.procurement.clarification.utils.JsonUtil;
import java.time.LocalDateTime;
import java.util.UUID;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.core.convert.ConversionService;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.internal.verification.VerificationModeFactory.times;

public class EnquiryServiceImplTest {
    private final static String ENQUIRY_ID = "44e128a5-ac7a-4c9a-be4c-224b6bf81b20";
    private final static LocalDateTime LOCAL_DATE_TIME = LocalDateTime.now();
    private final static AuthorDto AUTHOR_DTO = new AuthorDto("Author_name", "author_id");
    private final static String TITLE = "title";
    private final static String DESCRIPTION = "description";
    private final static String ANSWER = "anybody answer";
    private final static String RELATED_ID = "related_id";
    private final static String RELATED_LOT = "related_lot";
    private final static String THREAD_ID = "thread_id";
    private static final String OCID = "ocds-213czf-000-00001";

    private final EnquiryPeriodRepository periodRepository = mock(EnquiryPeriodRepository.class);
    private final EnquiryRepository enquiryRepository = mock(EnquiryRepository.class);
    private final ConversionService conversionService = mock(ConversionService.class);

    private final EnquiryServiceImpl enquiryService = new EnquiryServiceImpl(enquiryRepository, periodRepository, new
        JsonUtil
        (new ObjectMapper()), conversionService);

    @Test
    @DisplayName("save Enquiry")
    void saveEnquiryTest() {
        EnquiryDto enquiryDto = new EnquiryDto(ENQUIRY_ID,
                                               LOCAL_DATE_TIME,
                                               AUTHOR_DTO,
                                               TITLE,
                                               DESCRIPTION,
                                               ANSWER,
                                               LOCAL_DATE_TIME,
                                               RELATED_ID,
                                               RELATED_LOT,
                                               THREAD_ID);

        DataDto dataDto = new DataDto(OCID, enquiryDto);

        EnquiryEntity enquiryEntity = new EnquiryEntity();
        enquiryEntity.setCpId(OCID);
        enquiryEntity.setEnquiryId(UUID.randomUUID());
        enquiryEntity.setIsAnswered(true);
        enquiryEntity.setJsonData("json");

        EnquiryPeriodEntity enquiryPeriodEntity = new EnquiryPeriodEntity();
        enquiryPeriodEntity.setCpId(OCID);
        enquiryPeriodEntity.setStartDate(LocalDateTime.now()
                                                      .minusDays(2));
        enquiryPeriodEntity.setEndDate(LocalDateTime.now()
                                                    .plusDays(2));

        when(conversionService.convert(dataDto, EnquiryEntity.class)).thenReturn(enquiryEntity);

        EnquiryServiceImpl spy = spy(enquiryService);
        doNothing().when(spy)
                   .checkPeriod(LocalDateTime.now(), OCID);
        when(periodRepository.getByOcId(OCID)).thenReturn(enquiryPeriodEntity);
        enquiryService.saveEnquiry(dataDto);

        verify(enquiryRepository, times(1)).save(enquiryEntity);
    }

    @Test
    @DisplayName("Check Period")
    void checkPeriodTest() {

        when(periodRepository.getByOcId(OCID)).thenThrow(new NullPointerException("nullPointer"));
        assertThrows(NullPointerException.class, () -> enquiryService.checkPeriod(LOCAL_DATE_TIME, OCID));
    }

    @Test
    @DisplayName("Check Period")
    void checkPeriodIncorrectDateTest() {
        EnquiryPeriodEntity enquiryPeriodEntity = new EnquiryPeriodEntity();
        enquiryPeriodEntity.setCpId(OCID);
        enquiryPeriodEntity.setStartDate(LocalDateTime.now()
                                                      .minusDays(3));
        enquiryPeriodEntity.setEndDate(LocalDateTime.now()
                                                    .minusDays(2));

        when(periodRepository.getByOcId(OCID)).thenReturn(enquiryPeriodEntity);
        assertThrows(ErrorInsertException.class, () -> enquiryService.checkPeriod(LOCAL_DATE_TIME, OCID));
    }

}

File: src/test/kotlin/com/procurement/clarification/application/service/EnquiryServiceImplTest.kt
package com.procurement.clarification.application.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import com.procurement.clarification.application.repository.enquiry.EnquiryRepository
import com.procurement.clarification.application.repository.enquiry.model.EnquiryEntity
import com.procurement.clarification.application.repository.period.model.PeriodEntity
import com.procurement.clarification.domain.extension.format
import com.procurement.clarification.domain.model.Cpid
import com.procurement.clarification.domain.model.Ocid
import com.procurement.clarification.domain.model.enums.Stage
import com.procurement.clarification.domain.model.token.Token
import com.procurement.clarification.exception.ErrorException
import com.procurement.clarification.infrastructure.api.ApiVersion
import com.procurement.clarification.infrastructure.api.v1.CommandMessage
import com.procurement.clarification.infrastructure.api.v1.CommandTypeV1
import com.procurement.clarification.infrastructure.api.v1.Context
import com.procurement.clarification.infrastructure.bind.api.CommandId
import com.procurement.clarification.infrastructure.bind.configuration
import com.procurement.clarification.infrastructure.handler.v1.model.response.CheckAnswerRs
import com.procurement.clarification.infrastructure.service.JacksonJsonTransform
import com.procurement.clarification.json.loadJson
import com.procurement.clarification.lib.functional.asSuccess
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.format.ResolverStyle
import java.util.*

internal class EnquiryServiceImplTest {

    lateinit var generationService: GenerationService
    lateinit var enquiryRepository: EnquiryRepository
    lateinit var periodService: PeriodService
    lateinit var enquiryService: EnquiryService
    lateinit var transform: Transform

    companion object {
        private val CPID = Cpid.tryCreateOrNull("ocds-t1s2t3-MD-1565251033096")!!
        private val OCID = Ocid.tryCreateOrNull("ocds-b3wdp1-MD-1581509539187-EV-1581509653044")!!
        private val TOKEN = UUID.randomUUID()

        private const val FORMAT_PATTERN = "uuuu-MM-dd'T'HH:mm:ss'Z'"
        private val FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern(FORMAT_PATTERN)
            .withResolverStyle(ResolverStyle.STRICT)
        private val DATE = LocalDateTime.parse("2019-02-10T08:49:55Z", FORMATTER)
    }

    @Nested
    inner class CheckEnquiries {

        @BeforeEach
        fun init() {
            generationService = mock()
            enquiryRepository = mock()
            periodService = mock()
            enquiryService = EnquiryServiceImpl(generationService, enquiryRepository, periodService)

            val objectMapper = ObjectMapper().apply { configuration() }
            transform = JacksonJsonTransform(objectMapper)
        }

        @Test
        fun wrongId_exception() {
            val cm = getCommandMessage(id = "wrongId")
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actualException = assertThrows<ErrorException> {
                enquiryService.checkAnswer(cm)
            }
            val expectedCode = "00.01"
            assertEquals(expectedCode, actualException.code)
        }

        @Test
        fun wrongToken_exception() {
            val cm = getCommandMessage(id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3", token = UUID.randomUUID())
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actualException = assertThrows<ErrorException> {
                enquiryService.checkAnswer(cm)
            }
            val expectedCode = "20.06"
            assertEquals(expectedCode, actualException.code)
        }

        @Test
        fun alreadyAnswered_exception() {
            val cm = getCommandMessage(id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3")
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json"), isAnswered = true))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actualException = assertThrows<ErrorException> {
                enquiryService.checkAnswer(cm)
            }
            val expectedCode = "02.01"
            assertEquals(expectedCode, actualException.code)
        }

        @Test
        fun stageTPEndDatePrecedes_exception() {
            val cm = getCommandMessage(id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3", data = loadJson("json/service/answer/check/request_pre_qualification.json"))
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(actual.setUnsuspended, false)
        }

        @Test
        fun stageFEEndDatePrecedes_exception() {
            val cm = getCommandMessage(
                id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3",
                data = loadJson("json/service/answer/check/request_pre_qualification.json"),
                stage = Stage.FE
            )
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(actual.setUnsuspended, false)
        }

        @Test
        fun stagePCEndDatePrecedes_exception() {
            val cm = getCommandMessage(
                id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3",
                data = loadJson("json/service/answer/check/request_tender_period.json"),
                stage = Stage.PC
            )
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(actual.setUnsuspended, false)
        }

        @Test
        fun stageEVEndDatePrecedes_exception() {
            val cm = getCommandMessage(
                id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3",
                data = loadJson("json/service/answer/check/request_tender_period.json"),
                stage = Stage.EV
            )
            val enquiries = listOf(getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")))
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            whenever(periodService.getPeriodEntity(CPID, OCID)).thenReturn(PeriodEntity(CPID, OCID, "", DATE, DATE.plusDays(2)))

            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(actual.setUnsuspended, false)
        }

        @Test
        fun otherEnquiriesNotAnswered_exception() {
            val cm = getCommandMessage(
                id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3",
                data = loadJson("json/service/answer/check/request_tender_period.json"),
                stage = Stage.EV
            )
            val enquiries = listOf(
                getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")),
                getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data_1.json"), isAnswered = false)
                                      )
            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            whenever(periodService.getPeriodEntity(CPID, OCID)).thenReturn(PeriodEntity(CPID, OCID, "", DATE, DATE.minusDays(2)))

            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(actual.setUnsuspended, false)
        }

        @Test
        fun otherEnquiriesAnswered_exception() {
            val cm = getCommandMessage(
                id = "a63155e0-0320-11eb-b8f6-2d3c13365ab3",
                data = loadJson("json/service/answer/check/request_tender_period.json"),
                stage = Stage.EV
            )
            val enquiries = listOf(
                getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data.json")),
                getEnquiry(jsonData = loadJson("json/service/answer/check/enquiry_json_data_1.json"), isAnswered = true)
            )

            whenever(enquiryRepository.findBy(CPID, OCID)).thenReturn(enquiries.asSuccess())
            whenever(periodService.getPeriodEntity(CPID, OCID)).thenReturn(PeriodEntity(CPID, OCID, "", DATE, DATE.minusDays(2)))

            val actual = enquiryService.checkAnswer(cm).data as CheckAnswerRs

            assertEquals(true, actual.setUnsuspended)
        }

        private fun getCommandMessage(
            id: String,
            data: String = "{}",
            stage: Stage = Stage.TP,
            token: Token = TOKEN
        ) = CommandMessage(
            id = CommandId.invoke("id"),
            command = CommandTypeV1.CHECK_ANSWER,
            context = Context(
                cpid = CPID.underlying,
                ocid = OCID.underlying,
                token = token.toString(),
                startDate = DATE.format(),
                id = id,
                stage = stage.key,
                owner = null,
                country = null,
                endDate = null,
                access = null,
                language = null,
                operationId = null,
                operationType = null,
                phase = null,
                pmd = null,
                prevStage = null,
                processType = null,
                requestId = null,
                setExtendedPeriod = null,
            ),
            version = ApiVersion(0, 0, 0),
            data = transform.tryParse(data).get
        )

        private fun getEnquiry(
            token: Token = TOKEN,
            isAnswered: Boolean = false,
            jsonData: String
        ) = EnquiryEntity(
            cpid = CPID,
            ocid = OCID,
            token = token,
            isAnswered = isAnswered,
            jsonData = jsonData
        )
    }
}